source/dotnet/Library/AdaptiveCards/ParseContext.cs

```diff
@@ -0,0 +1,131 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+
+namespace AdaptiveCards
+{
+    static class ParseContext
+    {
+        public static void Initialize()
+        {
+            ElementIDs = new Dictionary<string, List<AdaptiveInternalID>>();
+            IDStack = new Stack<Tuple<string, AdaptiveInternalID, bool>>();
+        }
+
+        public enum ContextType { Element, Action };
+
+        public static ContextType Type { get; set; }
+
+        private static IDictionary<string, List<AdaptiveInternalID>> ElementIDs = new Dictionary<string, List<AdaptiveInternalID>>();
+
+        private static Stack<Tuple<string, AdaptiveInternalID, bool>> IDStack = new Stack<Tuple<string, AdaptiveInternalID, bool>>();
+
+        // Push the provided state on to our ID stack
+        public static void PushElement(string idJsonProperty, AdaptiveInternalID internalId)
+        {
+            if (internalId.Equals(AdaptiveInternalID.Invalid))
+            {
+                throw new AdaptiveSerializationException($"Attemping to push an element on to the stack with an invalid ID");
+            }
+            IDStack.Push(new Tuple<string, AdaptiveInternalID, bool>(idJsonProperty, internalId, AdaptiveFallbackConverter.IsInFallback));
+        }
+
+        // Pop the last id off our stack and perform validation 
+        public static void PopElement()
```

Comment:
>PopElement [](start = 27, length = 10)

This makes my head reel.. Assuming we got good code coverage for this function in our unit's?

Reply:
yes. this is an implementation of what we have in the shared model. I can give you a rundown sometime if you're interested. it's needed to solve a pretty complex problem :)

---
In reply to: [278345702](https://github.com/Microsoft/AdaptiveCards/pull/2560#discussion_r278345702) [](ancestors = 278345702)

Reply:
Yes pls! and ideally point me to the test case that i can debug away to glory as well when we talk about this :) thanks!

---
In reply to: [278346705](https://github.com/Microsoft/AdaptiveCards/pull/2560#discussion_r278346705) [](ancestors = 278346705,278345702)